README.md
# SOAP

SOAP version 2.0, downloaded from 
[this page](http://astro.up.pt/resources/soap2/).
The original readme is reproduced below.




      ######          ###########            #####         ###########        #######
    #        #      #             #        #       #       #          #     #         #
   #               #               #     #           #     #           #               #
  #                #        ##     #    #             #    #           #               #
   #               #        ##     #    #             #    #           #              #
     #             #               #    #             #    #          #             #
        #          #               #    #             #    ###########            #
           #       #               #    #             #    #                    #
             #     #               #    # ########### #    #                  #
              #    #               #    #             #    #                #
             #     #               #    #             #    #               #
   #        #       #             #     #             #    #               #
     ######           ###########       #             #    #               ###########
 

   |                               #
   |                           #        #
   |                         #            #
   |                       #                #
   |                     #                    #
   | ###################                        #                       #################
   |                                              #                   #
   |                                                #               #
   | WWW.ASTRO.UP.PT/RESOURCES/SOAP2                  #           #
   | Version 1.0 - Xavier Dumusque                      #       #
   | xdumusque@cfa.harvard.edu                              #
   |_____________________________________________________________________________________
   

#########################################
INTRODUCTION
#########################################

SOAP 2.0 is a code that estimates the effects of active regions, spots or plages, on radial-velocity and photometric measurements. SOAP 2.0 gives access to the radial-velocity, the bisector span, the full width at half maximum variations as defined with the cross-correlation technique and optimized to reproduce the HARPS (HARPS-N) observations. The photometric variation induced by active regions (flux at 5293 Angstrom) is also returned by SOAP 2.0.

The code is published in Dumusque et al. (2014, ApJ, 796, 132) and you are advised to look at this paper and reference therein for more information about the code. To summarize, the code is based on SOAP (Boise et al. 2012, A&A, 545, 109) with some modifications to include the effect of the inhibition of convection in magnetic regions due to strong local magnetic fields. This effect has been shown to be significant for stars similar to the Sun for which plages are dominating (Meunier et al 2010, A&A, 512, 39). To include this effect, we used observed spectra of the Sun taken with the Kitt Peak Fourrier Transform Spectrograph (FTS). Including this inhibition of the convective blueshift effect using solar spectra allows reproducing better the observations of slow rotators (see Dumusque 2014, ApJ, 796, 133).


#########################################
INSTALLATION
#########################################

The SOAP 2.0 code is written in C, with a python interface for plotting the results.
The code has been tested on python 2.6 and 2.7 and though it is expected to work properly on more recent versions, the user is advised to use this version in case of troubles.

The following standard python libraries are required:
- numpy 1.6.2 or higher
- scipy 0.11.0 or higher
- matplotlib 1.1.1 or higher
- pyfits 3.3 or higher

The following C libraries are required:
- gsl

You can download the latest version of SOAP 2.0 here:
www.astro.up.pt/ressources/soap2

Once downloaded, unzip the SOAP_2.zip file
$ unzip SOAP_2.zip

To run SOAP 2.0, you first have to compile the C code to be usable by your version of python.
To do so go on to the StarSpot folder and compile the C code using the setup.py file. Then copy the created file in the StarSpot folder:
$ cd SOAP_2/StarSpot                     (where SOAP_PATH is where you have installed SOAP 2.0)
$ python setup.py build                     (this create the wrapped file in build/lib.***) 
$ cp build/lib.****/starspot.so .           (where **** corresponds to your operating system)

You might have a warning "Using deprecated NumPy API...” when compiling the C file using the command “python setup.py build”, however this will not affect SOAP 2.


#########################################
RUNNING SOAP 2.0
#########################################

CONFIGURATION:
The configuration of SOAP 2.0 is done using the config.cfg file. Inside this ascii file, you can configure the resolution of the grid used for the simulation (grid) and the resolution of the active region circumference (nrho). The instrumental resolution can also be changed.  It is of course possible to configure the stellar properties (inclination, rotational period, radius..) and the active regions properties (active region type (spot, plage), longitude, latitude, size). Finally, it is possible to configure at which phase the output is generated. Everything is commented in the config.cfg file and it should be easy to configure it yourself.

Up to four active regions, spots or plages, can be created with SOAP 2.0, however it is easy to configure the program to accept more active regions.

Once the config.cfg file is configured to your needs, run the soap2.py file with the command  (you must be in the folder where you installed SOAP 2.0):

$ python soap2.py
or
$ ipython soap2.py (for the one that have installed ipython (much more flexible than python))


DISPLAY:
To work with python in interactive mode, which display interactive figures, you should configure the correct backend for you operating system. This can be done by opening the matplotlibrc file in the main SOAP_2 folder and change the value of the backend variable. For MacOSX, choose MacOSX, for Linux, choose belong GTKAgg, TkAGG, WX, WXAgg, QtAgg, Qt4Agg and test if it works (see here for more information http://matplotlib.org/faq/usage_faq.html#what-is-a-backend). You can check in your original matplotlibrc file what is the value for the backend, as this one should be correctly configured (this original file should be in the matplotlib directory that can be find in the python site-package folder: /Library/Python/2.7/site-packages/matplotlib-1.3.1-py2.7-macosx-10.9-intel.egg/matplotlib/mpl-data for my MacOSX installation)

If you are in interactive mode two plots appear. The first plot shows the two CCF derived from the quiet photosphere spectrum and from the spot spectrum of the Kitt Peak FTS solar observations. The second plot shows as a function of phase the effect in flux, radial-velocity (RV), bisector span (Bis Span) and full width at half maximum (Fwhm) of the disc-integrated CCF. Note that for each observable, the flux effect (only due to the contrast difference of the active regions), the convective blueshift effect (only due to the inhibition of the convective blueshift inside active regions) and the combined effect are returned.

If you are not in interactive mode (Agg backend), the figures are not displayed.

Note that if the wrong interactive backend is selected, python might crash (not always the case) because it cannot find the backend, and therefore no output will be generated. If you cannot find any interactive backend that works, you should select the Agg backend. You will not be able to have interactive figures, but they will be saved in .pdf format anyway.


OUTPUTS:
The following files will be created in a specific folder that can be found in the ‘outputs’ directory. The name of this folder depends on the active region configuration that you chose and will contain:
- A .pdf file for each generated figure
- a .rdb file that gives the value of all the observables for all calculated phase
- a ‘fits’ folder where .fits files for each phase is created. The headers of these files contain all the information about the configuration used and the calculated observables. The data of the fits files contain the disc-integrated CCF from which the RV, Bis Span and Fwhm are calculated.


CODE RAPIDITY:
The rapidity of the code can be increased by reducing the number of phase at which the outputs are calculated. In addition the efficiency of the code can be improved by reducing  the stellar grid resolution (GRID = 300) or the spot circumference resolution (NRHO = 20). I would not select a grid resolution smaller than 100 and a circumference resolution smaller than 10. For big active regions, the decrease in resolution will not influence strongly the observable variations, but be careful for small active regions. No serious test have been done to check the effect of small resolution and I would advise to always compare the results with the standard resolution (GRID = 300, NRHO = 20) to be sure that the observed variations are induced by active regions and not by numerical imprecision.


#########################################
TROUBLESHOOTINGS
#########################################

Did you make sure you are using python 2.6 or 2.7, and have installed the libraries specified above?

If you do not have one of the python libraries, you can install them that way:
$ sudo easy_install library_name (where library_name should be replaced by numpy, scipy...).
Another way is to use pip (https://pypi.python.org/pypi/pip)

If you do not have the C gsl library, you can install it easily on MacOSX with MacPort (https://www.macports.org/install.php). Once MacPort is installed, just type in a terminal the following command:
$ sudo port install gsl
Under Linux you can for example use yum to install gsl. Under Linux you will also need the gsl development package:
$ sudo yum install gsl
$ sudo yum install gsl-devel


If compiling the C code does not work because it can not find the numpy/arrayobject.h file
fatal error: ‘numpy/arrayobject.h’ file not found,
you have to manually update the path to where the file is. Open the setup.py file an update the  python_include_dir variable. For MacOSX, this should be something similar to:
‘/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/numpy/core/include’
or
‘/Library/Python/2.7/site-packages/numpy/core/include/’

Similar problem can happen for the gsl/gsl_errno.h file
‘gsl/gsl_errno.h' file not found.
In this case, update the  gsl_include_dir variable

If it cannot find the gsl libraries, it can also return an error message
‘library not found for -lgslcblas’.
In this case, update the gsl_lib_dir variable.

If still you cannot install and/or give the correct path to your libraries, please go and see you system administrator that will help you. This is not a problem of SOAP 2 and I will not reply to library problem requests.


#########################################
CONTACT
#########################################

Xavier Dumusque

xdumusque@cfa.harvard.edu

Harvard Smithsonian Center for Astrophysics
60 Garden Street
02138 Cambridge, MA
USA





CHANGELOG
- Original SOAP2.0 from webpage


StarSpot/starspot.h
/******************************************************************************/
/*** file:   starspot.h                                                     ***/
/*** author: X. Bonfils                                                     ***/
/***         Centro de Astronomia e Astrofisica da Universidade de Lisboa   ***/
/***                                                                        ***/
/*** version: 0.1 2006/08/29                                                ***/
/******************************************************************************/

#ifndef __PHASM_H__
#define __PHASM_H__

#include <stdlib.h>
#include <math.h>

#define DEBUG   0
#define VERBOSE 0
#define pi      3.14159265358

extern
void gauss_bis(double *vrad_ccf, double *ccf, double *err, int n1, double *mod, double *para, double *depth, double *bis, int len_depth);

extern 
void itot(double v, double i, double limba1, double limba2, double modif_bis_quad, double modif_bis_lin, double modif_bis_cte, int grid, 
          double *vrad_ccf, double *intensity_ccf, double v_interval, int n_v, int n, 
          double *f_star, double *sum_star);

extern
void starmap(double v, double i, double limba1, double limba2, int grid,
	  double **Fmap, double **Vmap);

extern 
void spot_init(double s, double longitude, double latitude, double inclination, 
	       int nrho, double **xyz);

extern 
void spot_phase(double **xyz, double inclination, int nrho, double psi, 
		double **xyz2);
		
extern
int spot_area(double **xlylzl, int nrho, int grid, int *iminy, int *iminz, 
	       int *imaxy, int *imaxz);
	       
extern
void spot_scan(double v, double i, double limba1, double limba2, double modif_bis_quad, double modif_bis_lin, double modif_bis_cte, int grid, 
	      double *vrad_ccf, double *intensity_ccf, double *intensity_ccf_spot, double v_interval, int n_v, int n, 
          double s, double longitude, double phase, double latitude, 
	      int iminy, int iminz, int imaxy, 
	      int imaxz, double *f_spot_flux, double *f_spot_bconv, double *f_spot_tot, double *sum_spot,
          int magn_feature_type, int T_star, int T_diff_spot);
	      
extern
void spot_inverse_rotation(double *xyz, double longitude, double latitude, 
			   double inclination, double phase, double *xiyizi);

extern
void spot_scan_npsi(double **xyz, int nrho, double *psi, int npsi, double v, 
                   double inclination, double limba1, double limba2, double modif_bis_quad, double modif_bis_lin, double modif_bis_cte, int grid, 
				   double *vrad_ccf, double *intensity_ccf, double *intensity_ccf_spot, double v_interval, int n_v, int n, 
				   double s, double longitude, double latitude, 
	      		   double **f_spot_flux, double **f_spot_bconv, double **f_spot_tot, double *sum_spot,
                   int magn_feature_type, int T_star, int T_diff_spot);

#endif /*__PHASM_H__*/


soap2.py
#!/usr/bin/env python

import numpy, os, sys
import pyfits
from pylab import *
import ConfigParser
import shutil
import string
sys.path.append('StarSpot/')
import PyStarSpot_soap2 as StSp

######################################################################
###           Functions            ###################################
######################################################################

# To read rdb file
def read_rdb(filename):
    
    f = open(filename, 'r')
    data = f.readlines()
    f.close()
    
    z=0
    while data[z][:2] == '# ' or data[z][:2] == ' #':
        z += 1

    key = string.split(data[z+0][:-1],'\t')
    output = {}
    for i in range(len(key)): output[key[i]] = []
    
    for line in data[z+2:]:
        qq = string.split(line[:-1],'\t')
        for i in range(len(key)):
            try: value = float(qq[i])
            except ValueError: value = qq[i]
            output[key[i]].append(value)

    return output

def write_rdb(filename,data,keys,format):
    
    f = open(filename, 'w')
        
    head1 = string.join(keys,'\t')
    head2 = ''
    for i in head1:
        if i=='\t': head2 = head2+'\t'
        else: head2 = head2+'-'

    f.write(head1+'\n')
    f.write(head2+'\n')
    
    if len(data.values()) > 0:
        for i in range(len(data.values()[0])):
            line = []
            for j in keys: line.append(data[j][i])
            f.write(format % tuple(line))
                            
    f.close()

####################################################################################
#                    INITITIALIZATION                     ##########################
####################################################################################

config = ConfigParser.ConfigParser()
config.read("config.cfg")

GRID              = int(config.get('main','grid' ))
NRHO              = int(config.get('main','nrho' ))
INST_RESO         = int(config.get('main','instrument_reso' ))
RAD_Sun           = int(config.get('star','radius_sun' ))
RAD               = float(config.get('star','radius' )) * RAD_Sun
PROT              = float(config.get('star','prot' ))
STAR              = StSp.Star(prot           = PROT,\
                              vrot           = (2.*pi*RAD)/(PROT*86400.),\
                              incl           = float(config.get('star','I')),\
                              limba1         = float(config.get('star','limb1')),\
                              limba2         = float(config.get('star','limb2')),\
                              psi            = float(config.get('star','psi')),\
                              rad_sun        = RAD_Sun,\
                              rad            = RAD,\
                              Temp           = int(config.get('star','Tstar')),\
                              Temp_diff_spot = int(config.get('star','Tdiff_spot')))

######################################################################
###           Read solar CCF       ###################################
######################################################################

if STAR.vrot < 10.:
    file1 = 'solar_CCFs/CCF_solar_spectrum_G2_FTS_reso_not_evenly_sampled_in_freq.rdb'
else:
    file1 = 'solar_CCFs/CCF_solar_spectrum_G2_FTS_reso_not_evenly_sampled_in_freq_extra_large_low_resolution.rdb'

data_ccf = read_rdb(file1)
rv_ccf = array(data_ccf['vrad'])
rv_ccf_magn_region = array(data_ccf['vrad'])
rv_ccf_sampling = rv_ccf[1]-rv_ccf[0]
intensity_ccf = array(data_ccf['CCF'])
rv_ccf_magn_region = array(data_ccf['vrad'])
intensity_ccf_magn_region = array(data_ccf['CCF_spot'])

####################################################################################
#                    INITITIALIZATION                     ##########################
####################################################################################

CCF               = StSp.Ccf(rv        = rv_ccf,\
                             intensity = intensity_ccf,\
                             width     = max(rv_ccf),\
                             step      = rv_ccf_sampling,\
                             star      = STAR)
               
CCF_active_region = StSp.Ccf(rv        = rv_ccf_magn_region,\
                             intensity = intensity_ccf_magn_region,\
                             width     = max(rv_ccf_magn_region),\
                             step      = rv_ccf_sampling,\
                             star      = STAR)

active_region1    = StSp.Active_region(check              = int(config.get('active_regions','check1')),\
                                       long               = float(config.get('active_regions','long1')),\
                                       lat                = float(config.get('active_regions','lat1')),\
                                       size               = float(config.get('active_regions','size1')),\
                                       active_region_type = int(config.get('active_regions','act_reg_type1')))

active_region2    = StSp.Active_region(check              = int(config.get('active_regions','check2')),\
                                       long               = float(config.get('active_regions','long2')),\
                                       lat                = float(config.get('active_regions','lat2')),\
                                       size               = float(config.get('active_regions','size2')),\
                                       active_region_type = int(config.get('active_regions','act_reg_type2')))

active_region3    = StSp.Active_region(check              = int(config.get('active_regions','check3')),\
                                       long               = float(config.get('active_regions','long3')),\
                                       lat                = float(config.get('active_regions','lat3')),\
                                       size               = float(config.get('active_regions','size3')),\
                                       active_region_type = int(config.get('active_regions','act_reg_type3')))

active_region4    = StSp.Active_region(check              = int(config.get('active_regions','check4')),\
                                       long               = float(config.get('active_regions','long4')),\
                                       lat                = float(config.get('active_regions','lat4')),\
                                       size               = float(config.get('active_regions','size4')),\
                                       active_region_type = int(config.get('active_regions','act_reg_type4')))

ACTIVE_REGIONS    = [active_region1,active_region2,active_region3,active_region4]

#calculate the visibility of the active regions
active_region1.calc_maps(STAR, GRID, NRHO)
active_region2.calc_maps(STAR, GRID, NRHO)
active_region3.calc_maps(STAR, GRID, NRHO)
active_region4.calc_maps(STAR, GRID, NRHO)

# Phase at which the signal we be generated
phase_step  = float(config.get('output','ph_step'))
phase_in    = config.get('output','ph_in')

# Read the phase in the file given in the "phase_in" argument in "config.cfg"
if phase_in!='None':
    f   = open(phase_in)
    PSI = array([float(l.split()[0]) for l in f.readlines()])
    f.close()
# Otherwise create a evenly sampled phase between 0 and 1 with the step given
# by the "ph_step" argument in "config.cfg"
elif phase_step>0.:
    PSI = arange(0.,1.,phase_step)
else:
    print
    print "WARNING: put a positive phase step ('ph_step' argument in 'config.cfg') or give a file name that contain the phases at which you want to compute the model ('phase_in' argument in 'config.cfg')"
    print


##########################################
#Calculate photometry and spectroscopy
##########################################

# Test if at least one active region has been activated in the config.cfg file
try:
    print
    print "****************************************************"
    print "Estimating the photometric and spectroscopic effects"
    print "****************************************************"
    print
    FLUXstar_quiet, CCFstar_quiet, flux, CCFstar_flux, CCFstar_bconv, CCFstar_tot, rv_flux, rv_bconv, rv_tot, span_flux, span_bconv, span_tot, fwhm_flux, fwhm_bconv, fwhm_tot, depth_flux, depth_bconv, depth_tot =\
                            StSp.Calculate_activity_signal(NRHO, GRID, STAR, PSI, CCF, CCF_active_region, ACTIVE_REGIONS,INST_RESO)

    # Calculates where rv_flux = rv_bconv, which corresponds to the phases where the active region is not visible
    index_equal_rv = where((rv_flux-rv_bconv)==0)[0]
    if len(index_equal_rv) != 0:
        zero_velocity = rv_flux[index_equal_rv][0] # velocity when the spot is not visible
        rv_flux  -= zero_velocity # Put the velocity when the spot is not visible to 0
        rv_bconv -= zero_velocity # Put the velocity when the spot is not visible to 0
        rv_tot   -= zero_velocity # Put the velocity when the spot is not visible to 0

    ########################################################################
    ################ Write to file #########################################
    ########################################################################

    print
    print "****************************************************"
    print "Writing to file"
    print "****************************************************"
    print
    CCF_folder_outputs = 'outputs/CCF_PROT=%.2f_i=%.2f_lon=(%.1f,%.1f,%.1f,%.1f)_lat=(%.1f,%.1f,%.1f,%.1f)_size=(%.4f,%.4f,%.4f,%.4f)/' % (STAR.prot,STAR.incl,active_region1.long,active_region2.long,active_region3.long,active_region4.long,active_region1.lat,active_region2.lat,active_region3.lat,active_region4.lat,active_region1.s,active_region2.s,active_region3.s,active_region4.s)
    CCF_folder_fits = CCF_folder_outputs + 'fits/'

    if os.path.exists(CCF_folder_outputs):
        shutil.rmtree(CCF_folder_outputs)
        os.mkdir(CCF_folder_outputs)
    else:
        os.mkdir(CCF_folder_outputs)

    if os.path.exists(CCF_folder_fits):
        shutil.rmtree(CCF_folder_fits)
        os.mkdir(CCF_folder_fits)
    else:
        os.mkdir(CCF_folder_fits)

    filename = "CCF_PROT=%.2f_i=%.2f_lon=(%.1f,%.1f,%.1f,%.1f)_lat=(%.1f,%.1f,%.1f,%.1f)_size=(%.4f,%.4f,%.4f,%.4f)" % (STAR.prot,STAR.incl,active_region1.long,active_region2.long,active_region3.long,active_region4.long,active_region1.lat,active_region2.lat,active_region3.lat,active_region4.lat,active_region1.s,active_region2.s,active_region3.s,active_region4.s)

    # Writing to fits file
    for i in range(len(PSI)):
        hdu = pyfits.PrimaryHDU(CCFstar_tot[i])
        hdulist = pyfits.HDUList([hdu])
        hdulist[0].header.set('STEP',CCF.step,'CCF Step [km/s]')
        hdulist[0].header.set('GRID',GRID,'Grid resolution')
        hdulist[0].header.set('NRHO',NRHO,'Spot circonference resolution')
        hdulist[0].header.set('RESO',NRHO,'Instrument resolution')
        hdulist[0].header.set('RADSUN',RAD_Sun,'Radius of the sun [km]')
        hdulist[0].header.set('RAD',STAR.rad,'Stellar radius [Rsun]')
        hdulist[0].header.set('PROT',STAR.prot,'Stellar Rotation Period [day]')
        hdulist[0].header.set('INCL',STAR.incl,'Stellar Inclination [degree]')
        hdulist[0].header.set('TSTAR',STAR.Temp,'Stellar effective temperature [K]')
        hdulist[0].header.set('TSPOTDIF',STAR.Temp_diff_spot,'Temp diff between photosphere and spot')
        hdulist[0].header.set('LIMBA1',STAR.limba1,'Stellar linear limb darkening coefficient')
        hdulist[0].header.set('LIMBA2',STAR.limba2,'Stellar linear limb darkening coefficient')
        hdulist[0].header.set('VROT',STAR.vrot,'Stellar Rotation [km/s] (calculated)')
        hdulist[0].header.set('TYPE1',active_region1.active_region_type_str,'Active region 1 type')
        hdulist[0].header.set('LONG1',active_region1.long,'Active region 1 longitude [degree]')
        hdulist[0].header.set('LAT1',active_region1.lat,'Active region 1 latitude [degree]')
        hdulist[0].header.set('SIZE1',active_region1.s,'Active region size [fraction of stellar radius]')
        hdulist[0].header.set('TYPE2',active_region2.active_region_type_str,'Active region 2 type')
        hdulist[0].header.set('LONG2',active_region2.long,'Active region 2 longitude [degree]')
        hdulist[0].header.set('LAT2',active_region2.lat,'Active region 2 latitude [degree]')
        hdulist[0].header.set('SIZE2',active_region2.s,'Active region size [fraction of stellar radius]')
        hdulist[0].header.set('TYPE3',active_region3.active_region_type_str,'Active region 3 type')
        hdulist[0].header.set('LONG3',active_region3.long,'Active region 3 longitude [degree]')
        hdulist[0].header.set('LAT3',active_region3.lat,'Active region 3 latitude [degree]')
        hdulist[0].header.set('SIZE3',active_region3.s,'Active region size [fraction of stellar radius]')
        hdulist[0].header.set('TYPE4',active_region4.active_region_type_str,'Active region 4 type')
        hdulist[0].header.set('LONG4',active_region4.long,'Active region 4 longitude [degree]')
        hdulist[0].header.set('LAT4',active_region4.lat,'Active region 4 latitude [degree]')
        hdulist[0].header.set('SIZE4',active_region4.s,'Active region size [fraction of stellar radius]')
        hdulist[0].header.set('RV_F',rv_flux[i],'RV for flux effect [km/s] (measured)')
        hdulist[0].header.set('SPAN_F',span_flux[i],'BIS for flux effect [km/s] (measured)')
        hdulist[0].header.set('FWHM_F',fwhm_flux[i],'FWHM for flux effect [km/s] (measured)')
        hdulist[0].header.set('RV_BC',rv_bconv[i],'RV for conv blue [km/s] (measured)')
        hdulist[0].header.set('SPAN_BC',span_bconv[i],'BIS for conv blue [km/s] (measured)')
        hdulist[0].header.set('FWHM_BC',fwhm_bconv[i],'FWHM for conv blue [km/s] (measured)')
        hdulist[0].header.set('RV',rv_tot[i],'RV [km/s] (measured)')
        hdulist[0].header.set('SPAN',span_tot[i],'BIS [km/s] (measured)')
        hdulist[0].header.set('FWHM',fwhm_tot[i],'FWHM [km/s] (measured)')
        hdulist[0].header.set('FLUX',flux[i],'Norm. Flux (measured)')
        hdulist[0].header.set('PSI',PSI[i],'Phase [0-1]')
        hdulist.writeto(CCF_folder_fits+filename+"_PSI=%.3f.fits" % PSI[i])

    # Writing to rdb file
    keys = ["Phase","Flux","RV_tot","Bis_Span_tot","Fwhm_tot","RV_flux","Bis_Span_flux","Fwhm_flux","RV_conv_blue","Bis_Span_conv_blue","Fwhm_conv_blue"]
    value = [PSI,flux,rv_tot,span_tot,fwhm_tot,rv_flux,span_flux,fwhm_flux,rv_bconv,span_bconv,fwhm_bconv]
    data = dict([(keys[i],value[i]) for i in arange(len(value))])
    format = "%f\t"*(len(value)-1) + "%f\n"
    write_rdb(CCF_folder_outputs+filename+'.rdb',data,keys,format)

    ########################################################################
    ################## Figures #############################################
    ########################################################################

    # Plot the solar CCFs used in the simulation
    figure(0)
    title('solar CCFs')
    plot(rv_ccf,intensity_ccf,color='b',lw=3,label='Quiet photosphere')
    plot(rv_ccf_magn_region,intensity_ccf_magn_region,color='r',ls='--',lw=3,label='Spot')
    ylim(0.4,1.05)
    xlim(-20,20)
    ylabel('Normalized flux')
    xlabel('RV [km/s]')
    legend(loc=4)
    subplots_adjust(top=0.93,left=0.1,right=0.96,bottom=0.13)
    savefig(CCF_folder_outputs+'FTS_solar_CCFs.pdf')

    majorFormatter = ScalarFormatter()
    majorFormatter.set_powerlimits((-8,8))
    majorFormatter.set_useOffset(0)

    # Plot the Flux, RV, BIS SPAN and FWHM variations induced by the active regions defined in "config.cfg"
    figure(1,[16,9])
    title('')
    subplot(511)
    plot(PSI,flux,color='r')
    ylabel('Norm. Flux',size=16)
    ax=gca()
    setp(ax.get_xticklabels(), visible=False)
    ax.yaxis.set_major_formatter(majorFormatter)
    subplot(512,sharex=ax)
    plot(PSI,rv_flux*1000,color='b',label='flux')
    plot(PSI,rv_bconv*1000,color='g',label='conv. blue.')
    plot(PSI,rv_tot*1000,color='r',label='tot')
    legend()
    ylabel('RV [m/s]',size=16)
    ax2=gca()
    setp(ax2.get_xticklabels(), visible=False)
    subplot(513,sharex=ax)
    plot(PSI,span_flux*1000,color='b')
    plot(PSI,span_bconv*1000,color='g')
    plot(PSI,span_tot*1000,color='r')
    ylabel('Bis Span [m/s]',size=16)
    ax3=gca()
    setp(ax3.get_xticklabels(), visible=False)
    subplot(514,sharex=ax)
    plot(PSI,fwhm_flux*1000,color='b',label='flux')
    plot(PSI,fwhm_bconv*1000,color='g',label='conv. blue.')
    plot(PSI,fwhm_tot*1000,color='r',label='tot')
    ylabel('Fwhm [m/s]',size=16)
    ax = gca()
    ax.yaxis.set_major_formatter(majorFormatter)
    xlabel('Phase',size=16)
    subplots_adjust(top=0.98,left=0.07,right=0.98,bottom=0.07,hspace=0.15)

    text1 =   '{0:19} = {1:<8d}, {2:23} = {3:<8d}\n'.format('Grid reso', GRID, 'Circumference reso', NRHO) + \
              '{0:19} = {1:<8d}, {2:23} = {3:<8d}\n'.format('Instr reso', INST_RESO, 'Radius of the sun [km]', RAD_Sun) + \
              '{0:19} = {1:<8.2f}, {2:23} = {3:<8.2f}\n'.format('Star Rot Period [d]', STAR.prot, 'Stellar radius [Rsun]', STAR.rad/RAD_Sun) + \
              '{0:19} = {1:<8.1f}, {2:23} = {3:<8.2f}\n'.format('Star incli', STAR.incl, 'Stellar vsini [m/s]', STAR.vrot) + \
              '{0:19} = {1:<8d}, {2:23} = {3:<8d}\n'.format('Stellar Teff [K]', STAR.Temp, 'Tdiff spot-photo [K]', STAR.Temp_diff_spot) + \
              '{0:19} = {1:<8.3f}, {2:23} = {3:<8.3f}'.format('Limb-dark lin', STAR.limba1, 'Limb-dark quad', STAR.limba2)

    text2 = 'Act Reg {0}: {1:<6}, lon = {2:<5.1f}, lat = {3:<4.1f}, size = {4:<5.3f} [Rsun] (= {5:.2f}%)\n'.format('1',active_region1.active_region_type_str,active_region1.long,active_region1.lat,active_region1.s,active_region1.s**2/2.*100) +\
            'Act Reg {0}: {1:<6}, lon = {2:<5.1f}, lat = {3:<4.1f}, size = {4:<5.3f} [Rsun] (= {5:.2f}%)\n'.format('2',active_region2.active_region_type_str,active_region2.long,active_region2.lat,active_region2.s,active_region2.s**2/2.*100) +\
            'Act Reg {0}: {1:<6}, lon = {2:<5.1f}, lat = {3:<4.1f}, size = {4:<5.3f} [Rsun] (= {5:.2f}%)\n'.format('3',active_region3.active_region_type_str,active_region3.long,active_region3.lat,active_region3.s,active_region3.s**2/2.*100) +\
            'Act Reg {0}: {1:<6}, lon = {2:<5.1f}, lat = {3:<4.1f}, size = {4:<5.3f} [Rsun] (= {5:.2f}%)'.format('4',active_region4.active_region_type_str,active_region4.long,active_region4.lat,active_region4.s,active_region4.s**2/2.*100)

    figtext(0.015,0.03,text1, name='Bitstream Vera Sans Mono',size=13) #name='Courier' very important because monotype font (all the characters have the same space)
    figtext(0.485,0.076,text2, name='Bitstream Vera Sans Mono',size=13) #name='Courier' very important because monotype font (all the characters have the same space)

    savefig(CCF_folder_outputs+filename+'.pdf')
#
## If no active region has been activated in the config.cfg file
except IndexError:
    print
    print 'ERROR: There is no active region selected for the star, please select at least one in the config.cfg file'
    print

show()